go-controller/pkg/clustermanager/networkconnect/hybrid_connect_subnet_allocator.go

```diff
@@ -0,0 +1,291 @@
+package networkconnect
+
+import (
+	"errors"
+	"fmt"
+	"net"
+	"sync"
+
+	utilnet "k8s.io/utils/net"
+
+	"github.com/ovn-org/ovn-kubernetes/go-controller/pkg/clustermanager/node"
+	"github.com/ovn-org/ovn-kubernetes/go-controller/pkg/config"
+)
+
+var (
+	p2pIPV4SubnetMask = 31
+	p2pIPV6SubnetMask = 127
+)
+
+// HybridConnectSubnetAllocator provides hybrid allocation for network connect subnets:
+//   - Layer3 networks: Each gets a full layer3NetworkPrefix block (e.g., /24)
+//   - Layer2 networks: Pooled allocation - multiple Layer2 networks share layer3NetworkPrefix blocks,
+//     with each Layer2 network getting a /31 (IPv4) or /127 (IPv6) from the shared pool
+//
+// This allocator uses the node.SubnetAllocator to allocate subnets underneath using the layer3NetworkPrefix.
+//
+// We run one instance of this allocator per CNC.
+type HybridConnectSubnetAllocator interface {
+	// AddNetworkRange initializes the allocator with the overall CIDR range and network prefix
+	AddNetworkRange(network *net.IPNet, networkPrefix int) error
+
+	// AllocateLayer3Subnet allocates network subnets for the layer3 network owner (could be both IPv4 and IPv6 in dual-stack)
+	AllocateLayer3Subnet(owner string) ([]*net.IPNet, error)
+
+	// AllocateLayer2Subnet allocates /31 (IPv4) and/or /127 (IPv6) for Layer2 networks from shared pools
+	// Returns subnets for all available address families in the cluster
+	AllocateLayer2Subnet(owner string) ([]*net.IPNet, error)
+
+	// MarkAllocatedSubnets marks multiple subnets as already allocated (for existing allocations)
+	MarkAllocatedSubnets(owner string, subnets ...*net.IPNet) error
+}
+
+// layer2SubnetPool manages dual-stack networkPrefix blocks subdivided into Layer2 subnets
+// Each pool can contain both IPv4 and IPv6 allocations
+type layer2SubnetPool struct {
+	v4PoolCIDR *net.IPNet              // IPv4 networkPrefix block (nil if not available)
+	v6PoolCIDR *net.IPNet              // IPv6 networkPrefix block (nil if not available)
+	allocator  node.SubnetAllocator    // Dual-stack allocator: subdivides v4 pool into /31s and v6 pool into /127s
+	allocated  map[string][]*net.IPNet // Track what's allocated from this pool (can be v4, v6, or both)
+}
+
+// hybridConnectSubnetAllocator implements HybridConnectSubnetAllocator
+type hybridConnectSubnetAllocator struct {
+	sync.RWMutex
```

Comment:
can you add a comment saying which fields/functions it protects?

Reply:
argh actually seems like I need to have a mutex after all. I now remember why I had all those convoluted caches... its to tend to the corner case scenario of - I release a layer2 network - I need to check if the entire layer3 pool is empty or not or whether its used by other owners. If empty, we want to release it as well and if not we want to keep it as is. So for that I need some caches since now we have multiple ranges per l2 allocator and i need to know which range of l3 the l2 got allocated out of.
I'll add a comment detailing this.